File: 03_pqtl_and_signal_attribution/run_cystatinc_cond_coloc.py
#!/usr/bin/env python3
# 胱抑素 C 第三信号对主共定位的影响：条件化/区域剔除敏感性
# 1) 从 1000G VCF 计算 rs13329240/rs7165675 与全部 LD 面板变异的 r（EUR 无关）
# 2) 近似条件化 ITPKA eQTL 与湿疹 GWAS → 重跑 coloc.abf（R 脚本完成）
# 3) 区域剔除敏感性：剔除 chr15:41.15-41.35Mb 后重跑
import pysam, numpy as np, pandas as pd

DIR = "/workspace/coloc_panel"
OUT = "/workspace/fable_fix/pqtl"
univ = pd.read_csv(f"{DIR}/snp_universe.csv").dropna(subset=["rsid"]).reset_index(drop=True)
ld_snps = pd.read_csv(f"{DIR}/ld_eur_snps.csv")["rsid"].tolist()

# 目标变异在 universe 中的键
targets = {}
for v in ["rs13329240", "rs7165675"]:
    r = univ[univ.rsid == v].iloc[0]
    targets[v] = (int(r.position), r.ref, r.alt)

eur = set(open(f"{DIR}/eur_unrel.txt").read().split())
vcf = pysam.VariantFile(f"{DIR}/kgp_chr15_window.vcf")
keep = [s for s in vcf.header.samples if s in eur]

# 提取目标变异剂量（定向到 universe ref→alt 尺度）
dos_t = {}
for rec in vcf:
    if len(rec.alts or []) != 1:
        continue
    ref, alt = rec.ref, rec.alts[0]
    if len(ref) != 1 or len(alt) != 1:
        continue
    for v, (pos, uref, ualt) in targets.items():
        if rec.pos != pos:
            continue
        if (ref, alt) == (uref, ualt):
            flip = False
        elif (ref, alt) == (ualt, uref):
            flip = True
        else:
            continue
        gt = np.array([rec.samples[s]["GT"] for s in keep], dtype=np.float32)
        if (gt < 0).any():
            continue
        d = gt.sum(axis=1)
        dos_t[v] = (2 - d) if flip else d
        print(f"{v}: pos={pos} flip={flip} EUR AF={d.mean()/2:.4f}")

# 面板变异剂量（复用 make_ld 逻辑，仅取 LD 面板内变异）
ld_set = set(ld_snps)
key2idx = {}
for i, r in univ.iterrows():
    if r.rsid in ld_set:
        key2idx[(int(r.position), r.ref, r.alt)] = r.rsid
vcf.close()
vcf = pysam.VariantFile(f"{DIR}/kgp_chr15_window.vcf")  # 重开（非索引流 reset 不可靠）
G = {}
for rec in vcf:
    if len(rec.alts or []) != 1:
        continue
    ref, alt = rec.ref, rec.alts[0]
    if len(ref) != 1 or len(alt) != 1:
        continue
    k1 = (rec.pos, ref, alt); k2 = (rec.pos, alt, ref)
    if k1 in key2idx:
        rsid, flip = key2idx[k1], False
    elif k2 in key2idx:
        rsid, flip = key2idx[k2], True
    else:
        continue
    gt = np.array([rec.samples[s]["GT"] for s in keep], dtype=np.float32)
    if (gt < 0).any():
        continue
    d = gt.sum(axis=1)
    G[rsid] = (2 - d) if flip else d

# 计算 r
rows = []
for v, dt in dos_t.items():
    dt_c = dt - dt.mean()
    for rsid in ld_snps:
        if rsid not in G:
            continue
        g = G[rsid]; g_c = g - g.mean()
        denom = np.sqrt((g_c**2).sum() * (dt_c**2).sum())
        if denom == 0:
            continue
        r = float((g_c * dt_c).sum() / denom)
        rows.append({"cond_variant": v, "rsid": rsid, "r": r})
out = pd.DataFrame(rows)
out.to_csv(f"{OUT}/cystatinc_cluster_ld.csv", index=False)
print(out.groupby("cond_variant")["r"].apply(lambda s: s.abs().max()))
print("saved cystatinc_cluster_ld.csv", out.shape)

File: 03_pqtl_and_signal_attribution/run_cystatinc_cond_coloc.R
#!/usr/bin/env Rscript
# 胱抑素 C 第三信号条件化后的主共定位稳健性
suppressMessages({library(coloc); library(data.table)})
DIR <- "/workspace/coloc_panel"; OUT <- "/workspace/fable_fix/pqtl"

eq <- fread(file.path(DIR, "eqtlgen_ITPKA.csv"))
gw <- fread(file.path(DIR, "gwas_finngen_L12_DERMATITISECZEMA.tsv"), header=FALSE)
setnames(gw, c("chromosome","position","ref","alt","rsids","nearest_genes","pval","mlogp","beta","sebeta","af_alt","af_alt_cases","af_alt_controls"))
univ <- fread(file.path(DIR, "snp_universe.csv"))
ld  <- fread(file.path(OUT, "cystatinc_cluster_ld.csv"))

# GWAS 展开 rsids（多 rsid 分号分隔取第一个匹配 universe 的）
gw[, rsid := sapply(strsplit(rsids, ";"), function(x) x[1])]
m <- merge(eq, gw, by.x="rsid", by.y="rsid", suffixes=c("_eq","_gw"))
m <- merge(m, univ[, .(rsid, maf)], by="rsid")
m <- m[maf > 1e-6 & maf < 0.999999]
cat("merged nsnps:", nrow(m), "\n")

run_coloc <- function(b1, se1, b2, se2, maf, N1, N2, label) {
  d1 <- list(beta=b1, varbeta=se1^2, N=N1, type="quant", MAF=maf, snp=seq_along(b1))
  d2 <- list(beta=b2, varbeta=se2^2, N=N2, type="cc", MAF=maf, snp=seq_along(b2))
  r <- coloc.abf(d1, d2)
  cat(sprintf("%-42s H3=%.4f H4=%.4f\n", label, r$summary["PP.H3.abf"], r$summary["PP.H4.abf"]))
  invisible(r$summary)
}

res <- list()
res[["unconditioned"]] <- run_coloc(m$beta_eq, m$se, m$beta_gw, m$sebeta, m$maf, 30744, 500348, "unconditioned eQTL x eczema")

# 近似条件化：z_{j|c} = (z_j - r z_c)/sqrt(1-r^2)；转回 beta 尺度用 se_c = se/sqrt(1-r^2)
for (cv in c("rs13329240","rs7165675")) {
  rr <- ld[cond_variant == cv, .(rsid, r)]
  mm <- merge(m, rr, by="rsid")
  mm <- mm[abs(r) < 0.99]
  # eQTL 侧
  zc_eq <- mm[mm$rsid == cv, beta_eq/se]; if (length(zc_eq)==0) { # 条件变异不在 merged 集则用其最近代理
    prox <- mm[which.max(abs(r))][1]
    zc_eq <- prox$beta_eq/prox$se; cat(cv, "eQTL 侧用代理", prox$rsid, "r=", prox$r, "\n") }
  # GWAS 侧
  zc_gw <- mm[mm$rsid == cv, beta_gw/sebeta]; if (length(zc_gw)==0) {
    prox <- mm[which.max(abs(r))][1]
    zc_gw <- prox$beta_gw/prox$sebeta; cat(cv, "GWAS 侧用代理", prox$rsid, "r=", prox$r, "\n") }
  f <- sqrt(1 - mm$r^2)
  b1c <- (mm$beta_eq - mm$r * zc_eq * mm$se) / f^2   # beta 尺度条件化
  se1c <- mm$se / f
  b2c <- (mm$beta_gw - mm$r * zc_gw * mm$sebeta) / f^2
  se2c <- mm$sebeta / f
  res[[paste0("cond_", cv)]] <- run_coloc(b1c, se1c, b2c, se2c, mm$maf, 30744, 500348, paste0("conditioned on ", cv, " (n=", nrow(mm), ")"))
}

# 区域剔除：剔除胱抑素 C 峰区 41.15-41.35 Mb（hg38）
pos <- univ$position[match(m$rsid, univ$rsid)]
keep <- !(pos >= 41150000 & pos <= 41350000)
res[["drop_region"]] <- run_coloc(m$beta_eq[keep], m$se[keep], m$beta_gw[keep], m$sebeta[keep], m$maf[keep], 30744, 500348, paste0("region 41.15-41.35Mb dropped (n=", sum(keep), ")"))

out <- data.frame(analysis=names(res),
                  H3=sapply(res, `[`, "PP.H3.abf"), H4=sapply(res, `[`, "PP.H4.abf"))
fwrite(out, file.path(OUT, "cystatinc_cond_coloc_results.csv"))
print(out)
